syntax = "proto3";

package serialization;

message Type {
  uint32 category = 1;
  oneof o {
    uint32 index = 2;
    uint64 value = 3;
  }
}

message TypeSystem {
  message ParameterType {
    message Parameter {
      string name = 1;
      Type type = 2;
    }
    repeated Parameter parameters = 1;
  }

  repeated ParameterType parameters = 1;
  repeated Type pointers = 2;
  repeated Type buffer_pointers = 3;
  message ArrayType {
    Type type = 1;
    uint64 length = 2;
  }
  repeated ArrayType arrays = 4;
  repeated Type slices  = 5;

  message FunctionType {
    uint32 parameters = 1;
    repeated Type return_types = 2;
  }
  repeated FunctionType functions = 6;

  message EnumType {
    map<string, uint64> enumerator = 1;
  }
  repeated EnumType enums = 7;
}
